addons/html_editor/static/src/editor/core/history_plugin.js

```diff
@@ -2,37 +2,585 @@
 
 import { registry } from "@web/core/registry";
 import { Plugin } from "../plugin";
+import { setSelection } from "../utils/selection";
+import { serializeSelection } from "../utils/serialize";
 
 export class HistoryPlugin extends Plugin {
     static name = "history";
-    static shared = ["enableObserver", "disableObserver"];
+    static dependencies = ["dom"];
+
+    // @todo @phoenix usefull for the collaboration plugin. See if we still want
+    // to handle the serialization through that property later.
+    isSerializable = false;
 
     setup() {
-        this.observer = new MutationObserver(() => this.handleDOMChange());
-        this.enableObserver();
+        this.registry.category("shortcuts").add("control+z", { command: "HISTORY_UNDO" });
+        this.registry.category("shortcuts").add("control+y", { command: "HISTORY_REDO" });
+        this.addDomListener(this.editable, "input", this.addStep);
+        this.addDomListener(this.editable, "keydown", this.stageSelection);
+        this.addDomListener(this.editable, "beforeinput", this.stageSelection);
+        this.observer = new MutationObserver(this.handleNewRecords.bind(this));
         this._cleanups.push(() => this.observer.disconnect());
     }
+    start() {
+        this.enableObserver();
+        this.reset();
+        this.stageFirstStep();
+        this.renderingClasses = new Set(
+            this.registry.category("history_rendering_classes").getAll().flat()
+        );
+    }
+    handleCommand(command, payload) {
+        switch (command) {
+            case "HISTORY_UNDO":
+                this.undo();
+                break;
+            case "HISTORY_REDO":
+                this.redo();
+                break;
+            case "ADD_STEP":
+                this.addStep();
+                break;
+            case "HISTORY_STAGE_SELECTION":
+                this.stageSelection();
```

Comment:
What does staging mean in this context? Same for `stageFirstStep`.

Reply:
Similar as in git, we add to the `currentStep`. We might revert it or change it before we commit it.